cpp/src/delivery_test.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include <proton/connection.hpp>
+#include <proton/connection_options.hpp>
+#include <proton/container.hpp>
+#include <proton/delivery.h>
+#include <proton/delivery.hpp>
+#include <proton/link.hpp>
+#include <proton/listen_handler.hpp>
+#include <proton/listener.hpp>
+#include <proton/message.hpp>
+#include <proton/message_id.hpp>
+#include <proton/messaging_handler.hpp>
+#include <proton/tracker.hpp>
+#include <proton/types.h>
+#include <proton/types.hpp>
+#include <proton/value.hpp>
+
+#include "proton/error_condition.hpp"
+#include "proton/internal/pn_unique_ptr.hpp"
+#include "proton/receiver_options.hpp"
+#include "proton/transport.hpp"
+#include "proton/work_queue.hpp"
+#include "test_bits.hpp"
+
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace {
+std::mutex m;
+std::condition_variable cv;
+bool listener_ready = false;
+int listener_port;
+} // namespace
+
+class test_recv : public proton::messaging_handler {
+  private:
+    class listener_ready_handler : public proton::listen_handler {
+        void on_open(proton::listener &l) PN_CPP_OVERRIDE {
+            {
+                std::lock_guard<std::mutex> lk(m);
+                listener_port = l.port();
+                listener_ready = true;
+            }
+            cv.notify_one();
+        }
+    };
+
+    std::string url;
+    proton::listener listener;
+    listener_ready_handler listen_handler;
+
+  public:
+    test_recv(const std::string &s) : url(s) {}
+
+    void on_container_start(proton::container &c) PN_CPP_OVERRIDE {
+        listener = c.listen(url, listen_handler);
+    }
+
+    void on_message(proton::delivery &d, proton::message &msg) PN_CPP_OVERRIDE {
+        d.receiver().close();
+        d.connection().close();
+        listener.stop();
```

Comment:
did you step through the code in debugger? I think you stop the container by doing these three things; does the ASSERT_EQUAL that's after this even run? 

since you've added `on_delivery_settle`, I think you should stop the receiver there; that should be the last event callback that gets called, unless I am mistaken

Reply:
It looks like the ASSERT_EQUAL that is right after those three lines is running but the ASSERT_EQUAL that is inside `on_delivery_settle` is not. I tried to fix that one but I am not able to understand when `on_delivery_settle` will be called, as on debugging `on_delivery_settle` is not even getting called. 

![image](https://user-images.githubusercontent.com/16265285/115284665-9696c280-a16a-11eb-8e2e-d0da699db8f1.png)

I tried to investigate more about `on_delivery_settle`, and I found an archived thread in the user mailing list related to it, I am not sure if that's related to the issue I am facing. 
https://www.mail-archive.com/users@qpid.apache.org/msg15078.html

Meanwhile investigating more.

Reply:
https://github.com/apache/qpid-proton/blob/8ddf5399013b02f1fd13bda3fee7886bd48a98be/cpp/src/messaging_adapter.cpp#L118-L144

The execution goes through line 133 there, so it misses the only call to `on_delivery_settle` in the entire codebase below in the else if branch. There's no easy way to get `on_delivery_settle` called that I can see, so please remove it again. Sorry, bad advice the first time.

BTW, good idea using the coverage to see what's happening. Do you have a debugger too? My first thought would be to go for a debugger; it does not give such a persuasive screenshot that you've provided, but on the other hand it actually tells you what's happening line after line...

Reply:
> The execution goes through line 133 there, so it misses the only call to `on_delivery_settle` in the entire codebase below in the else if branch.

Oh okay, got it! Removed it. Thanks!

>Do you have a debugger too? My first thought would be to go for a debugger; it does not give such a persuasive screenshot that you've provided, but on the other hand it actually tells you what's happening line after line...

No, I haven't used it. Will look into it.

Reply:
> No, I haven't used it. Will look into it.

OK, let me give you a few pointers. If  you already use an IDE, you should use something built-in into it. If you use e.g. VS Code, there should be a debugger for C++ as a plugin. It is good to use a debugger in a program you already know how to use. Learning a new IDE at the same time makes things hard. What IDE/editor are you using for C++?

Here's a tutorial for VS Code. Looking at it, I feel QT Creator was simpler ;P https://www.youtube.com/watch?v=G9gnSGKYIg4

When I had C++ at school, we used the QT Creator IDE which was quite nice, or Clion is ok, if you have a free licence from your school (which we had). Another possibility is Eclipse for C/C++ which @ganeshmurthy is probably still using.

There is a command-line debugger GDB (and LLDB, which looks essentially the same). Pretty much all the other graphical debuggers are simply graphical front ends for GDB. I do not recommend using GDB directly when you have another option.

Reply:
> If you already use an IDE, you should use something built-in into it. If you use e.g. VS Code, there should be a debugger for C++ as a plugin. It is good to use a debugger in a program you already know how to use. Learning a new IDE at the same time makes things hard. What IDE/editor are you using for C++?

I am using VS Code for C++.

> Here's a tutorial for VS Code. Looking at it, I feel QT Creator was simpler ;P https://www.youtube.com/watch?v=G9gnSGKYIg4

Thank you :) I will look into it.

Reply:
Is the VS Code debugger working for you?

Reply:
Yes, It's working but I tested it on a sample code only. Still figuring out how to use it for qpid proton.

Reply:
Here's what I'd do. `ctest -VV -N -R delivery_test` will give you command that ctest uses to run test. You'd need to write this into the VSCode launch configuration. Some of the env variables are not important, and some are. LD_LIBRARY_PATH is important, for example. Best to find a minimal command that will work (in commandline) and then transfer that to the launch configuration.

If you want to run an example, etc, it's the same. Find out the minimal command on a (fresh) commandline, without doing `source build/config.sh` and put that into VSC launch config.

Reply:
It's working now. Thanks, Jiri. It really helped.

Reply:
Awesome. As a further improvement you could look into, there's probably something like these .env files, https://github.com/microsoft/vscode-python/issues/544. The idea is that you don't have to specify env variables in the launch config, which is quite clumsy, but instead you can keep them in a more readable format in a file. But I guess there was not too many environment variables necessary. Probably just that `LD_LIBRARY_PATH` I mentioned, right?

If you want some more work now, and to practice debugging ;P you could have a look at https://issues.apache.org/jira/browse/PROTON-2104. (Unless Justin finds you something better, or it's already time to look into your main task.) If you recall your tests for the url class, there were some tests for IPv6, and they were passing. So, how it's possible that IPv6 connections don't work? I tried stepping through the code in a debugger some time ago, and I left a comment. You'd have to double-check if the observations are still valid. Use the step-into command to follow the program execution, and see where the url string breaks. Then, you can propose a solution, where to best apply a fix.

## Motivational cartoon, just for fun ;)

![image](https://user-images.githubusercontent.com/442720/115968329-aed65b00-a537-11eb-9997-65d291fece92.png)

Reply:
Thank you, ack :)